scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala

```diff
@@ -146,33 +147,34 @@ class BspClient(private val workingDir: Path,
         .map(r => r.getItems())
         .map(classes => {
           if (classes.size() == 0) throw NoMainClassFound("No main class found.")
-          classes.get(0)
+          if (classes.get(0).getClasses().size() == 0) throw NoMainClassFound("No main class found.")
+          classes.get(0).getClasses().get(0)
         })
         // Run
         .flatMap(mainClass => {
-          wrapTimeout(s"$id-run", bspServer.buildTargetRun({
-            val param = new RunParams(targetId)
-            param.setDataKind("scala-main-class")
-            param.setData(mainClass.getClasses().get(0))
-            param.setOriginId(s"$id-run")
-            param
-          }))
+          bspServer.jvmRunEnvironment(new JvmRunEnvironmentParams(java.util.List.of(targetId)))
+            .asScala.map((mainClass, _))
         })
-        .map(result => {
-          if (result.getStatusCode() == StatusCode.ERROR) throw FailedRunError(s"Failed run: $result.")
-          if (result.getStatusCode() == StatusCode.CANCELLED) throw FailedRunError(s"Cancelled run: $result")
-          result
+        .map({
+          case (mainClass, runSettings) if runSettings.getItems().isEmpty() => throw BspClient.FailedRunError("Could not find run settings.")
```

Comment:
The same goes here, refactor it to Either

Reply:
Done… :)